### PR TITLE
Create Single Strand for Listener

### LINKED ISSUE #91: HTTP Handler not synchronously processing incoming messages
### Description
The HTTP Handler is creating a boost strand for every individual connection request. This means that we will have one strand per HTTPServer. 

#### Expected Behavior
We have one strand per listener which is synchronously handling POST requests. 

#### Actual Behavior
We have one strand per HTTPServer.

### PR DESCRIPTION
**Pull request by samdai01**
### Description
Adds a strand member to Listener and creates a single strand for that individual listener's use

- Resolves #91

### CODE REVIEW

File: projects/remote_transceiver/inc/remote_transceiver.h [hunk 1]
```diff
@@ -157,7 +157,7 @@ class Listener : public std::enable_shared_from_this<Listener>
      * @param acceptor tcp::acceptor configured with desired host and port
      * @param db       SailbotDB instance - the Listener requires that it takes ownership of the db
      */
-    Listener(bio::io_context & io, tcp::acceptor acceptor, SailbotDB && db);
+    Listener(tcp::acceptor acceptor, SailbotDB && db, bio::strand<bio::io_context::executor_type> strand);
```
Comment (hhenry01) [inline projects/remote_transceiver/inc/remote_transceiver.h:160]: I think it would be cleaner to just do what's done in the [official example](https://github.com/boostorg/beast/blob/f8ff44d072803e8c193946818ee7d239c5834b10/example/http/server/async/http_server_async.cpp#L336). Meaning instead of taking an acceptor and strand as arguments, take an io_context and tcp::endpoint{address, port} and make a new strand and acceptor in the constructor.

File: projects/remote_transceiver/src/remote_transceiver.cpp [hunk 2]
```diff
@@ -62,12 +62,12 @@ HTTPServer::HTTPServer(tcp::socket socket, SailbotDB & db) : socket_(std::move(s
...
 
-Listener::Listener(bio::io_context & io, tcp::acceptor acceptor, SailbotDB && db)
-: io_(io), acceptor_(std::move(acceptor)), db_(std::move(db)){};
+Listener::Listener(tcp::acceptor acceptor, SailbotDB && db, bio::strand<bio::io_context::executor_type> strand)
+: acceptor_(std::move(acceptor)), db_(std::move(db)), strand_(std::move(strand)){};
 
 void Listener::run()
 {
-    acceptor_.async_accept(bio::make_strand(io_), [&](beast::error_code e, tcp::socket socket) {
+    acceptor_.async_accept(strand_, [&](beast::error_code e, tcp::socket socket) {
```
Comment (hhenry01) [inline projects/remote_transceiver/src/remote_transceiver.cpp:70]: Again, I think we should just do what's done in the [official example](https://github.com/boostorg/beast/blob/f8ff44d072803e8c193946818ee7d239c5834b10/example/http/server/async/http_server_async.cpp#L393) and make a new strand for each request. That way, if we need to use io_context in the future, we won't have to worry about synchronization after accepting a request.

**Review by hhenry01 — CHANGES_REQUESTED** [review]
It'll probably be easier for you to combine this PR with your other PR. That way you won't have to keep switching between branches to make changes.

File: projects/remote_transceiver/inc/remote_transceiver.h (hunk 1 above)
Comment (samdai01) [reply projects/remote_transceiver/inc/remote_transceiver.h]: modified to follow example and constructor now just takes an endpoint

File: projects/remote_transceiver/src/remote_transceiver.cpp (hunk 2 above)
Comment (samdai01) [reply projects/remote_transceiver/src/remote_transceiver.cpp]: listener now creates a new strand on new connection which is managed by main acceptor strand

File: projects/remote_transceiver/src/remote_transceiver.cpp [hunk 3]
```diff
@@ -62,8 +62,35 @@ HTTPServer::HTTPServer(tcp::socket socket, SailbotDB & db) : socket_(std::move(s
...
+        acceptor_.listen(bio::socket_base::max_listen_connections, ec);
+        if (ec) {
+            throw(ec);
+        }
+    } catch (beast::error_code ec) {
+        std::cerr << "Error: " << ec.message() << std::endl;
+    }
+};
 
 void Listener::run()
```
Comment (samdai01) [inline projects/remote_transceiver/src/remote_transceiver.cpp:95]: in the official example, the equivalent of our `run()` utilizes `beast::bind_front_handler()` and passes it a pointer to a helper function which calls `async_accept`. When I try this the tests hang, do you have any idea why?

**Review by samdai01 — COMMENTED** [review]
bind_front_handler not used?

File: projects/remote_transceiver/src/remote_transceiver.cpp (hunk 3 above)
Comment (hhenry01) [reply projects/remote_transceiver/src/remote_transceiver.cpp:95]: The general concept of a `bind()` function is to call a callback function respective to a specific instance of an object. In the example, `bind()` is used to associate the callback function `on_accept()` with the current instance of the `listener` class. Currently in remote_transceiver, a lambda function is used to accomplish the same thing (mostly). I personally like [this explanation of lambdas](https://learn.microsoft.com/en-us/cpp/cpp/lambda-expressions-in-cpp?view=msvc-170). In short:

1. Specifying `[&]` in the lambda means that it's capturing all variables in the current scope.
2. This means we are capturing the current instance of the `Listener` object and its member variables/functions.
3. This accomplishes the same thing as `bind()`, but with less indirection. The callback function is super short and only used by this one `acceptor_.async_accept()` line, so why write a new function somewhere else in the source file?

tl;dr they do the same thing.

Comment (hhenry01) [reply projects/remote_transceiver/src/remote_transceiver.cpp:95]: To be 100% equivalent, you can change the `[&]` to `[this]` or `[shared_from_this()]` like in `readReq()` [here](https://github.com/UBCSailbot/network_systems/blob/ed2b1d5529d482227ad230f5821aa0b054c3a2c8/projects/remote_transceiver/src/remote_transceiver.cpp#L87-L88). I didn't realize I captured the object instance in `run()` and `readReq()`, so please change them to whatever you like so long as it's consistent.

Comment (samdai01) [reply projects/remote_transceiver/src/remote_transceiver.cpp:95]: changed listener to use `this` instead of `&`. Tried using shared_from_this() but it doesn't let me. I'm thinking it may be because Listener is created using `std::unique_ptr` instead of `shared_ptr()`

File: projects/remote_transceiver/src/remote_transceiver.cpp [hunk 4]
```diff
@@ -62,19 +62,47 @@ HTTPServer::HTTPServer(tcp::socket socket, SailbotDB & db) : socket_(std::move(s
...
 {
-    acceptor_.async_accept(bio::make_strand(io_), [&](beast::error_code e, tcp::socket socket) {
+    std::shared_ptr<Listener> self = shared_from_this();
+    acceptor_.async_accept(bio::make_strand(io_), [self](beast::error_code e, tcp::socket socket) {
         if (!e) {
-            std::make_shared<HTTPServer>(std::move(socket), db_)->doAccept();
+            std::make_shared<HTTPServer>(std::move(socket), self->db_)->doAccept();
```
Comment (hhenry01) [inline projects/remote_transceiver/src/remote_transceiver.cpp:100]: Please change to `std::move(self->db_)`

**Review by hhenry01 — APPROVED** [review]
